changelog/dmd.shortened-method-constructor.dd

```diff
@@ -0,0 +1,15 @@
+Shortened method syntax can now be used in constructors
+
+This used to raise an error "cannot return expression from constructor", but it's now supported:
+
+---
+struct Number
+{
+    int x;
+
+    this(int x) => this.x = x;
+    this(float x) => this(cast(int) x);
+}
+---
+
+Postblits and destructors already supported shortened method syntax because they return `void`.
```

Comment:
This causes an error though:
```d
void f() => 5;
```
shortfn.d(1): Error: cannot return non-void from `void` function

Shouldn't `=> expr;` always mean `{ return expr; }`?

Reply:
> The ShortenedFunctionBody form implies a [return statement](https://dlang.org/spec/statement.html#ReturnStatement).

https://dlang.org/spec/function.html#function-bodies

Reply:
> This causes an error though:

That's because of a type mismatch, `~this() => 5;` also causes an error, but `~this() => free()` doesn't when `free()` returns `void`.

Reply:
> Shouldn't `=> expr;` always mean `{ return expr; }`?

Yes, I only altered the rewrite in the parser because of semantic's weird handling of constructor returns. The way it works is that from the user side, they return `void`, but secretely, they are rewritten to return `this` by `ref`. Either way, you'd expect `this(x) => this(f(x))` to work: whether they return `void` or `this`, they're both the same.

The reason it doesn't is because dmd's statement semantic sets the returned expression from each return statement to be `this`, so if there already were a `sideEffect()` expression in there, that would get lost. I could try to extract the side effects from return statements in constructors and rewrite them more smartly, but I figured this would do as well, since the user's illusion of constructors returning void remains uphold.

Reply:
As a future enhancement I think explicit `return voidFunc();` in constructors should work as well though.

Reply:
> you'd expect this(x) => this(f(x)) to work

OK, but:
```d
struct Number
{
    int x;
    this(int x) => this.x = x; // surely the assignment has type int
}
```

Reply:
That's true...